InternetStoreCHIReloaded/BLL/Services/ProductsService.cs

```diff
@@ -0,0 +1,219 @@
+                }
+            }
+            else
+            {
+                result.IsSuccessful = false;
+                result.Message = "Fields are not proper.";
+            }
+
+            return result;
+        }
+
+        public async Task<ServiceResult> DeleteProductAsync(int id)
+        {
+            bool success = await _productRepository.RemoveAsync(id);
+
+            var result = new ServiceResult() { IsSuccessful = success };
+
+            if (success)
+            {
+                result.Message = "Product delete successful!";
+            }
+            else
+            {
+                result.Message = "No changes were made.";
+            }
+
+            return result;
+        }
+
+        public async Task<ServiceResult<Product>> GetProductAsync(int id)
+        {
+            var product = await _productRepository.FindFirstOrDefaultAsync(pe => pe.Id == id);
+
+            var result = new ServiceResult<Product>();
+
+            if (product != null)
+            {
+                result.IsSuccessful = true;
+                result.Message = "Product retrieval success!";
+                result.Data = _mapper.Map<Product>(product);
+            }
+            else
+            {
+                result.IsSuccessful = false;
+                result.Message = "Product was not found.";
+            }
+
+            return result;
+        }
+
+        public async Task<ServiceResult<List<Product>>> GetProductsAsync()// when method with pagination will be implemented, this method should be erased
+        {
+            var products = await _productRepository.GetAllAsync();
+
+            var result = new ServiceResult<List<Product>>()
+            {
+                IsSuccessful = true,
+                Data = _mapper.Map<List<Product>>(products),
+                Message = "Products retrieval successful!"
+            };
+
+            return result;
+        }
+
+        public async Task<ServiceResult<List<Product>>> GetProductsFilteredAsync(
+            int pageSize,
+            int page,
+            ProductsFilter filter)
+        {
+            var result = new ServiceResult<List<Product>>();
+
+            if (filter.FromPrice != null && filter.ToPrice != null)
+                if (filter.FromPrice > filter.ToPrice)
+                {
+                    result.IsSuccessful = false;
+                    result.Message = "FromPrice can't be bigger then ToPrice";
+                    return result;
+                }
+
+            Expression<Func<ProductEntity, bool>> whereExpr = ep => true;
+
+            if (filter.CategoryId != null)
+                whereExpr = whereExpr.And((pe) => pe.CategoryId == filter.CategoryId);
+
+            if (filter.FromPrice != null)
+                whereExpr = whereExpr.And((pe) => pe.Price >= filter.FromPrice);
+
+            if (filter.ToPrice != null)
+                whereExpr = whereExpr.And((pe) => pe.Price <= filter.ToPrice);
```

Comment:
This logic is overcomplicated. 
We could store IQueryable variable like:
```
var productsQuery = _context.Products.AsQueryable();
```

Later we could add as many .Where() conditions as you wish, like:

```
if (minPrice != null)
{
  productsQuery = productsQuery.Where(i => i.Price >= minPrice.Value);
}
```

This code should be part of the DAL repository method.